src/pydicom/dataset.py

```diff
@@ -440,6 +443,49 @@ def __exit__(
         exc_tb: TracebackType | None,
     ) -> bool | None:
         """Method invoked on exit from a with statement."""
+        if exc_val is None or sys.version_info < (3, 11):
+            return None
+
+        # Find first data element or Sequence up the trace
+        tb = exc_val.__traceback__
+        all_frames = list(traceback.walk_tb(tb))
+        note = ""
+        new_elem = False
+        raw_elem = False
+        for frame, lineno in reversed(all_frames):
+            filename = Path(frame.f_code.co_filename).name
+            match (filename, frame.f_code.co_name):
+                case ("filewriter.py", _):
+                    elem = frame.f_locals.get("elem")
+                case ("dataelem.py", "convert_raw_data_element"):
+                    elem = frame.f_locals.get("raw")
+                    if elem:
+                        raw_elem = True
+                case ("dataelem.py" | "sequence.py" | "dataset.py", _):
+                    elem = frame.f_locals.get("self")
+                case _:
+                    elem = None
+            if elem is None:
+                continue
+            if elem is self:
+                break
+            if path := path_to(elem, self):
+                note = "Error occurred at " + path
+                if new_elem:
+                    note += "\n   with DataElement not yet assigned to Dataset"
+                if raw_elem:
+                    note += f"\n  Converting RawDataElement(vr='{elem.VR}', value={elem.value!r}) "
+                break
+            else:
+                new_elem = filename == "dataelem.py"
+
+        # Check if note already exists and don't repeat
+        # This is so both pydicom and user can use this context manager
+        if not hasattr(exc_val, "__notes__") or all(
+            n != note for n in exc_val.__notes__
+        ):
+            exc_val.add_note(note)
```

Comment:
If `elem` is `None` then an empty note gets added here.

Reply:
Fixed